Escape from uncaught exceptions for IDEs

EricssonXD opened the pull request:
<!-- Write down your pull request descriptions. -->

I faced the same issue as #314 and #1869, and I also understand that it is not an error with the dio package but an error with the Flutter SDK / VScode, where VScode does not recognise the exceptions as handled when using Future.catchError(). 

Despite it not being a problem of this package, I believe "fixing" it such that it avoids the above mentioned problem is still the correct approach, the following modification avoids this problem, and allows the example Error Handling code on pub.dev works with VScode.

### New Pull Request Checklist

- [x] I have read the [Documentation](https://pub.dev/documentation/dio/latest/)
- [x] I have searched for a similar pull request in the [project](https://github.com/cfug/dio/pulls) and found none
- [x] I have updated this branch with the latest `main` branch to avoid conflicts (via merge from master or rebase)
- [x] I have added the required tests to prove the fix/feature I'm adding
- [x] I have updated the documentation (if necessary)
- [x] I have run the tests without failures
- [x] I have updated the `CHANGELOG.md` in the corresponding package

### Additional context and info (if any)

<!-- Provide more context and info about the PR. -->

AlexV525 commented:
Could also help to identify if this is relevant fix for #1869?

EricssonXD commented:
> Could also help to identify if this is relevant fix for #1869?

I forgot to mention it in the description, this is also an issue that I came across which is the exact problem I ran into initially

EricssonXD commented:
> @EricssonXD Hi, thanks for taking the time to investigate the cause. Very much appreciated!
> 
> I was wonder if adding the paired `async/await` could be addressed by simply adding `return` to the `_dispatchRequest`. WDYT?

As the function is a parameter that gets passed to the interceptor, which expects a `void Function(RequestOptions options, RequestInterceptorHandler handler)`, it doesn't work. Unless we also modify the interceptor?

EricssonXD commented:
> @EricssonXD Hi, thanks for taking the time to investigate the cause. Very much appreciated!
> 
> I was wonder if adding the paired `async/await` could be addressed by simply adding `return` to the `_dispatchRequest`. WDYT?

Is there any reasons why you would like to avoid `async/await`? From what I see in how the code is used, I don't think there should be any difference in terms of performance. But I'm still new to Flutter, so correct me if I'm wrong.

AlexV525 commented:
> Is there any reasons why you would like to avoid `async/await`? From what I see in how the code is used, I don't think there should be any difference in terms of performance. But I'm still new to Flutter, so correct me if I'm wrong.

You may spot that we've already implemented `.then` and `.catchError` for the method. Theoretically, all exceptions should be caught here. Awaiting a future that is not void is not recommended.

Consider writing like this:
![image](https://github.com/cfug/dio/assets/15884415/33a07b1e-661c-415d-a4da-67ec423fc798)

EricssonXD commented:
> > Is there any reasons why you would like to avoid `async/await`? From what I see in how the code is used, I don't think there should be any difference in terms of performance. But I'm still new to Flutter, so correct me if I'm wrong.
> 
> You may spot that we've already implemented `.then` and `.catchError` for the method. Theoretically, all exceptions should be caught here. Awaiting a future that is not void is not recommended.
> 
> Consider writing like this: ![image](https://private-user-images.githubusercontent.com/15884415/299199208-33a07b1e-661c-415d-a4da-67ec423fc798.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYwODA5MjcsIm5iZiI6MTcwNjA4MDYyNywicGF0aCI6Ii8xNTg4NDQxNS8yOTkxOTkyMDgtMzNhMDdiMWUtNjYxYy00MTVkLWE0ZGEtNjdlYzQyM2ZjNzk4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTI0VDA3MTcwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlmZDlhYzJmYjkwM2VlMTE4NmEzNWZkMzk4ODZjODdjNjI1MjRkMTczM2M0MzVhMzQxYWZhZmZlMzk2Njc3N2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.lzz8nhLT_lCFuAYJ4cUrF4C74rZdlMXrXsrAotSRZwk)

That's actually what I wrote initially during debugging but I found out adding only the await keyword somehow made it work which lead to my proposed code. I have updated my code now, but I don't quite understand what do you mean by 
>Awaiting a future that is not void is not recommended.
Isn't this new code also awaiting a future that is not void? Or do you mean like we shouldn't await a value that is not void if we are not using it?

AlexV525 commented:
> > Awaiting a future that is not void is not recommended.
> 
> Isn't this new code also awaiting a future that is not void? Or do you mean like we shouldn't await a value that is not void if we are not using it?

Sorry, that's a typo. :(

Awaiting a void future is not recommended.

AlexV525 commented:
> That's actually what I wrote initially during debugging but I found out adding only the await keyword somehow made it work which lead to my proposed code. I have updated my code now, but I don't quite understand what do you mean by
> 
> > Awaiting a future that is not void is not recommended.
> 
> Isn't this new code also awaiting a future that is not void? Or do you mean like we shouldn't await a value that is not void if we are not using it?

What I mean is we shouldn't wait for a future that returns value but not use it eventually.

In general:
```dart
Future<String> foo() async {
  await Future(() {});
  return 'bar';
}

void main() async {
  // Good.
  final result = await foo();
  print(result);

  // Bad.
  await foo();
}
```

EricssonXD commented:
Ahhh I see, are there any more problems that I should address in my pull request?

AlexV525 commented on dio/CHANGELOG.md:
```diff
@@ -10,6 +10,7 @@ See the [Migration Guide][] for the complete breaking changes list.**
 - Fix cancellation for streamed responses and downloads when using `IOHttpClientAdapter`.
 - Fix receive progress for streamed responses and downloads when using `IOHttpClientAdapter`.
 - Support relative `baseUrl` on the Web platform.
+- Fix `Excpetions` getting treated as `Uncaught` during debugging
```
```suggestion
- Avoid fake uncaught exceptions during debugging with IDEs.
```

AlexV525 reviewed (approved):
One nit:

AlexV525 commented:
I'd consider the request as test exempt, as it doesn't affect any actual behavior. Waiting for other inputs.

AlexV525 commented:
@kuhnroyal @ueman Any inputs?

AlexV525 commented:
Going to merge this and apply inputs if any in the future.

kuhnroyal commented:
I think the fix for the root cause is out in Dart:master now. But this looks cleaner in any case.
We should try to avoid `then` & `catchError` in all new could just for readability.
